shared/actions/chat/inbox.js

```diff
@@ -2,141 +2,279 @@
-      const conv = incoming.chatInboxConversation.params.conv
-      const supersedesState: Constants.SupersedesState = _inboxConversationLocalToSupersedesState(conv)
-      const supersededByState: Constants.SupersededByState = _inboxConversationLocalToSupersededByState(conv)
-      const finalizedState: Constants.FinalizedState = _conversationLocalToFinalized(conv)
+    const supersededBy = _toSupersedeInfo(conversationIDKey, c.supersededBy || [])
+    if (supersededBy) {
+      yield put(Creators.updateSupersededByState(Map({[conversationIDKey]: supersededBy})))
+    }
 
-      if (supersedesState.count()) {
-        yield put(Creators.updateSupersedesState(supersedesState))
-      }
-      if (supersededByState.count()) {
-        yield put(Creators.updateSupersededByState(supersededByState))
-      }
-      if (finalizedState.count()) {
-        yield put(Creators.updateFinalizedState(finalizedState))
-      }
+    if (c.info.finalizeInfo) {
+      yield put(Creators.updateFinalizedState(Map({[conversationIDKey]: c.info.finalizeInfo})))
+    }
+  }
 
-      if (conversation) {
-        yield put(Creators.updateInbox(conversation))
-        const selectedConversation = yield select(Constants.getSelectedConversation)
-        if (selectedConversation === conversation.get('conversationIDKey')) {
-          // load validated selected
-          yield put(Creators.loadMoreMessages(selectedConversation, false))
-        }
-      }
+  const inboxState = _conversationLocalToInboxState(c)
+
+  if (inboxState) {
+    yield put(Creators.updateInbox(inboxState))
+
+    // inbox loaded so rekeyInfo is now clear
+    yield put(Creators.clearRekey(inboxState.get('conversationIDKey')))
+
+    // Try and load messages if the updated item is the selected one
+    const selectedConversation = yield select(Constants.getSelectedConversation)
+    if (selectedConversation === inboxState.get('conversationIDKey')) {
+      // load validated selected
+      yield put(Creators.loadMoreMessages(selectedConversation, false))
+    }
+  }
+}
+
+// Gui is showing boxed content, find some rows to unbox
+function * untrustedInboxVisible (action: Constants.UntrustedInboxVisible): SagaGenerator<any, any> {
+  const {conversationIDKey, rowsVisible} = action.payload
+  const inboxes = yield select(state => state.chat.get('inbox'))
+
+  const idx = inboxes.findIndex(inbox => inbox.conversationIDKey === conversationIDKey)
+  if (idx === -1) {
+    return
+  }
+
+  // Collect items to unbox
+  const total = rowsVisible * 2
+  const conversationIDKeys = inboxes.slice(idx, idx + total).map(i => i.state === 'untrusted' ? i.conversationIDKey : null).filter(Boolean).toArray()
+
+  if (conversationIDKeys.length) {
+    yield call(_unboxConversations, conversationIDKeys)
+  }
+}
+
+// Loads the trusted inbox segments
+function * _unboxConversations (conversationIDKeys: Array<Constants.ConversationIDKey>): Generator<any, any, any> {
```

Comment:
we could likely change the other one, lemme see how much code change it is. this one one is better as it streams the results back. The other one won't give you results until its all done